Sources/Moya/MoyaProvider+Internal.swift

```diff
@@ -123,22 +123,10 @@ public extension MoyaProvider {
                 }
             }
 
-            switch endpoint.sampleResponseClosure() {
-            case .networkResponse(let statusCode, let data):
-                let response = Moya.Response(statusCode: statusCode, data: data, request: request, response: nil)
-                let result = validate(response)
-                plugins.forEach { $0.didReceive(result, target: target) }
-                completion(result)
-            case .response(let customResponse, let data):
-                let response = Moya.Response(statusCode: customResponse.statusCode, data: data, request: request, response: customResponse)
-                let result = validate(response)
-                plugins.forEach { $0.didReceive(result, target: target) }
-                completion(result)
-            case .networkError(let error):
-                let error = MoyaError.underlying(error, nil)
-                plugins.forEach { $0.didReceive(.failure(error), target: target) }
-                completion(.failure(error))
-            }
+            let completeResult: MoyaResult = stubbedResult.map { Response(statusCode: $0.statusCode, data: $0.data, request: request, response: $0.response) }
+            let validatedResponse = completeResult.flatMap { validate($0) }
+            plugins.forEach { $0.didReceive(validatedResponse, target: target) }
```

Comment:
I can't see a good use case where the other plugin methods e.g. `didReceive` would be used on any stub plugin ...
Any examples @amaurydavid?

Reply:
I'm not sure I understand your point. `didReceived` is called whether the response is a real response or a stubbed response. Once we built the stubbed response and we are ready to use it, we first notify plugins about the incoming response.

Reply:
My point is that if a user implements a stub plugin, there any case where it would be useful to implement other method besides the stub one? e.g. did receive

Reply:
If we think in terms of "1 plugin = 1 role" I don't think so. But I guess we could also have a single plugin acting as a stubbing plugin + AccesTokenPlugin + NetworkActivityPlugin.
Do you ask because thinking about separating stubbing into a proper plugin type ? I've thought about this but did not find a way to use multiple plugin types in the `provider.plugins` property without allowing to use non pluginType objects in it.

Reply:
> Do you ask because thinking about separating stubbing into a proper plugin type ?

Exactly, that goes to the point that plugins are much more general and since stub is much specific, seems like a good idea. And defining a proper stub structure we can have just the specific things we need for stubbing.

Reply:
Also, I think this allows users to make mistakes that would be hard to find out ... example: If you define a stub together in a `Plugin1` and another member of your team in `Plugin2` for the same case of a target, you might get an error on your stub being different that you expect and it could be really hard to track down from where this is coming from ... if you don't know that maybe there is some defined in other place and there isn't a rule defined to which one would be used :)